src/handlers/issue/pre.ts

```diff
@@ -0,0 +1,33 @@
+import { extractImportantWords, countIncludedWords, upsertCommentToIssue, listIssuesForRepo } from "../../helpers";
+import { getBotContext } from "../../bindings";
+import { Payload } from "../../types";
+
+export const findDuplicateOne = async () => {
+  const { payload: _payload } = getBotContext();
+  const issue = (_payload as Payload).issue;
+
+  if (!issue?.body) return;
+  const importantWords = await extractImportantWords();
+  const wordCount = importantWords.length;
+  let fetchDone = false;
+  const perPage = 10;
+  let curPage = 1;
+  issue;
+
+  while (!fetchDone) {
+    const issues = await listIssuesForRepo("all", perPage, curPage, "created", "asc");
+    for (const iss of issues) {
+      if (iss.body && iss.number != issue.number && !iss.pull_request) {
+        const probability = (countIncludedWords(iss.body, importantWords) * 100.0) / wordCount;
+        if (probability > parseInt(process.env.SIMILARITY_THRESHOLD || "80")) {
```

Comment:
I'm concerned about the probability calculation strategy. 

We should be leveraging ChatGPT's ability to interpret synonyms. It appears that you're manually calculating this based on the frequency of specific words being used.

Reply:
I am not sure about this. Do you want ChatGPT to find all similar words from word list in the issues? Assuming there are 1000 past issues, then we need to send 1001 requests to OpenAI. Or do you want ChatGPT to make all possible words from important words list?

Reply:
Well, my original vision is something more like getting a batch of the issue specifications (let's say whatever is close to maxing out the context window of the LLM, starting from latest issue number) and asking GPT to compare the newly created issue with every other issue in that batch. Then if it thinks there is a duplicate, to post a comment back to the duplicate issue.

Reply:
> We could do things without a vector database in order to simplify things, here's how:

> A new issue is posted
> We ask ChatGPT to extract a word list of the most "important" (i.e. unique adjectives?) words (for example, when I wanted to find this issue, I searched for the term "duplicate")
> We search the repository for all issues with the important words
> We go from highest issue number (most recent) and read the specification.
> If >80% confidence, stop the search and link back to it with a warning saying that it's likely to be a duplicate.
> This approach might be a little more brittle (rate limits) but we won't have to worry about the database which should make implementation and maintenance much easier.

But I think we decided to measure similarity by rate of important words's appearance. Not asking ChatGPT to compare the newly created issue and other issues.

Reply:
Perhaps there was a misunderstanding but refer to [this explanation](https://github.com/ubiquity/ubiquibot/issues/488#issuecomment-1668970677) of my perspective.

1. I parse out the "important" words of the newly created issue e.g. "duplicate"
2. Then I search those important words on GitHub issues search
3. Which will return me a list of all the specifications of somewhat similar issues (i believe this returned seven issues, which all have the keyword "duplicate")
4. Use ChatGPT to interpret the meaning of all seven specifications and compare to the current created one to see if it is describing the same thing.

Reply:
Then what is the mechanism to measure similarity? @pavlovcik 
Search issues for the top three important words and we can just only ask ChatGPT if the searched issues are similar to the new one, not similarity but yes or no.

Reply:
I think I'm okay with boolean over percentage IF 
1. We can not figure out how to prompt it to return confidence percentage and 
2. Are leveraging chatGPT's ability to understand meaning vs trying to manually calculate a percentage, throwing out all the advanced interpretation capabilities of ChatGPT

Reply:
I will try to make prompts to figure out percentage.

Reply:
I tried this kind of prompt: "I have two github issues.\n%first%\nand\n%second%\nPlease give me the possibility of the 2 issues are the same content.\n Give me in number format and add % after the number.\nDo not tell other things since I only need the number."
It seems GPT returns me the possibility in %.
But the problem here is we are comparing found issues with newly created one and the OpenAI returns me 429 (request rate limit) error.
Are we going to use paid OpenAI API key?

Reply:
QA for byteballet/santa-bringyouwishes#75

Reply:
> Are we going to use paid OpenAI API key? 

We're going to try to get as far as we can without paying. 

When I searched in my previous example I shared above, there were only seven issues to compare, which is really not a lot

Reply:
I think API has only paid tier, no free.

Reply:
OpenAI Api has free plan.
![image](https://github.com/ubiquity/ubiquibot/assets/103924884/701d520d-0a91-419c-8a57-61d91128dad2)

Reply:
Those are rate limits. I think for [free trial](https://help.openai.com/en/articles/4936830-what-happens-after-i-use-my-free-tokens-or-the-3-months-is-up-in-the-free-trial) you get 5$ and that's all